diff --git a/phase/src/models/imputation_hmm.cpp b/phase/src/models/imputation_hmm.cpp
index 820e90e..3a35681 100644
--- a/phase/src/models/imputation_hmm.cpp
+++ b/phase/src/models/imputation_hmm.cpp
@@ -26,16 +26,16 @@
 #include <models/imputation_hmm.h>
 
 inline
-float horizontal_add (const __m256& a)
+float horizontal_add (const simde__m256& a)
 {
-    __m128 vlow = _mm256_castps256_ps128(a);
-    __m128 vhigh = _mm256_extractf128_ps(a, 1); // high 128
-   vlow = _mm_add_ps(vlow, vhigh);     // add the low 128
-   __m128 shuf = _mm_movehdup_ps(vlow);        // broadcast elements 3,1 to 2,0
-   __m128 sums = _mm_add_ps(vlow, shuf);
-   shuf = _mm_movehl_ps(shuf, sums); // high half -> low half
-   sums = _mm_add_ss(sums, shuf);    // (no wasted instructions, and all of them are the 4B minimum)
-   return _mm_cvtss_f32(sums);
+    simde__m128 vlow = simde_mm256_castps256_ps128(a);
+    simde__m128 vhigh = simde_mm256_extractf128_ps(a, 1); // high 128
+   vlow = simde_mm_add_ps(vlow, vhigh);     // add the low 128
+   simde__m128 shuf = simde_mm_movehdup_ps(vlow);        // broadcast elements 3,1 to 2,0
+   simde__m128 sums = simde_mm_add_ps(vlow, shuf);
+   shuf = simde_mm_movehl_ps(shuf, sums); // high half -> low half
+   sums = simde_mm_add_ss(sums, shuf);    // (no wasted instructions, and all of them are the 4B minimum)
+   return simde_mm_cvtss_f32(sums);
 }
 
 imputation_hmm::imputation_hmm(conditioning_set * _C) {
@@ -78,7 +78,7 @@ void imputation_hmm::computePosteriors(const std::vector < float > & HL, std::ve
 }
 
 void imputation_hmm::forward(std::vector < bool > & flat) {
-	const __m256i _vshift_count = _mm256_set_epi32(31,30,29,28,27,26,25,24);
+	const simde__m256i _vshift_count = simde_mm256_set_epi32(31,30,29,28,27,26,25,24);
 	const unsigned int nstates = C->n_states;
 	const unsigned int nstatesMD8 = (nstates / 8) * 8;
 
@@ -96,16 +96,16 @@ void imputation_hmm::forward(std::vector < bool > & flat) {
 			{
 				const float fact1 = C->t[l-1] / nstates;
 				const float fact2 = C->nt[l-1] / AlphaSum[l-1];
-				const __m256 _fact1 = _mm256_set1_ps(fact1);
-				const __m256 _fact2 = _mm256_set1_ps(fact2);
-				__m256 _sum = _mm256_set1_ps(0.0f);
+				const simde__m256 _fact1 = simde_mm256_set1_ps(fact1);
+				const simde__m256 _fact2 = simde_mm256_set1_ps(fact2);
+				simde__m256 _sum = simde_mm256_set1_ps(0.0f);
 				int k = 0;
 				for (; k < nstatesMD8; k += 8)
 				{
-					const __m256 _prob_prev = _mm256_load_ps(&Alpha[(l-1)*modK+k]);
-					const __m256 _prob_curr = _mm256_fmadd_ps(_prob_prev, _fact2, _fact1);
-					_sum = _mm256_add_ps(_sum, _prob_curr);
-					_mm256_store_ps(&Alpha[l*modK+k], _prob_curr);
+					const simde__m256 _prob_prev = simde_mm256_load_ps(&Alpha[(l-1)*modK+k]);
+					const simde__m256 _prob_curr = simde_mm256_fmadd_ps(_prob_prev, _fact2, _fact1);
+					_sum = simde_mm256_add_ps(_sum, _prob_curr);
+					simde_mm256_store_ps(&Alpha[l*modK+k], _prob_curr);
 				}
 				if (k) AlphaSum[l] = horizontal_add(_sum);
 				for (int offset = nstatesMD8; offset < nstates ; offset ++) {
@@ -117,22 +117,22 @@ void imputation_hmm::forward(std::vector < bool > & flat) {
 		else
 		{
 			const std::array<float,2> emit = {Emissions[2*C->polymorphic_sites[l]+0], Emissions[2*C->polymorphic_sites[l]+1]};
-			const __m256 _emit0 = _mm256_set1_ps(emit[0]);
-			const __m256 _emit1 = _mm256_set1_ps(emit[1]);
+			const simde__m256 _emit0 = simde_mm256_set1_ps(emit[0]);
+			const simde__m256 _emit1 = simde_mm256_set1_ps(emit[1]);
 			if (l == 0)
 			{
 				const float fact1 = 1.0f / nstates;
-				const __m256 _fact1 = _mm256_set1_ps(fact1);
-				__m256 _sum = _mm256_set1_ps(0.0f);
+				const simde__m256 _fact1 = simde_mm256_set1_ps(fact1);
+				simde__m256 _sum = simde_mm256_set1_ps(0.0f);
 				int k = 0;
 				for (; k < nstatesMD8; k += 8)
 				{
-					const __m256i _bcst = _mm256_set1_epi32((unsigned int )C->Hvar.getByte(l, k));
-					const __m256i _mask = _mm256_sllv_epi32(_bcst, _vshift_count);
-					const __m256 _emiss = _mm256_blendv_ps (_emit0, _emit1, _mm256_castsi256_ps(_mask));
-					const __m256 _prob_curr = _mm256_mul_ps(_emiss, _fact1);
-					_sum = _mm256_add_ps(_sum, _prob_curr);
-					_mm256_store_ps(&Alpha[l*modK+k], _prob_curr);
+					const simde__m256i _bcst = simde_mm256_set1_epi32((unsigned int )C->Hvar.getByte(l, k));
+					const simde__m256i _mask = simde_mm256_sllv_epi32(_bcst, _vshift_count);
+					const simde__m256 _emiss = simde_mm256_blendv_ps (_emit0, _emit1, simde_mm256_castsi256_ps(_mask));
+					const simde__m256 _prob_curr = simde_mm256_mul_ps(_emiss, _fact1);
+					_sum = simde_mm256_add_ps(_sum, _prob_curr);
+					simde_mm256_store_ps(&Alpha[l*modK+k], _prob_curr);
 				}
 				if (k) AlphaSum[l] = horizontal_add(_sum);
 				for (int offset = nstatesMD8; offset < nstates ; offset ++)
@@ -145,20 +145,20 @@ void imputation_hmm::forward(std::vector < bool > & flat) {
 			{
 				const float fact1 = C->t[l-1] / nstates;
 				const float fact2 = C->nt[l-1] / AlphaSum[l-1];//AlphaSum2[l-1];// AlphaSum[l-1];
-				const __m256 _fact1 = _mm256_set1_ps(fact1);
-				const __m256 _fact2 = _mm256_set1_ps(fact2);
-				__m256 _sum = _mm256_set1_ps(0.0f);
+				const simde__m256 _fact1 = simde_mm256_set1_ps(fact1);
+				const simde__m256 _fact2 = simde_mm256_set1_ps(fact2);
+				simde__m256 _sum = simde_mm256_set1_ps(0.0f);
 
 				int k = 0;
 				for (; k < nstatesMD8; k += 8)
 				{
-					const __m256i _mask = _mm256_sllv_epi32(_mm256_set1_epi32((unsigned int )C->Hvar.getByte(l, k)), _vshift_count);
-					const __m256 _emiss = _mm256_blendv_ps (_emit0, _emit1, _mm256_castsi256_ps(_mask));
-					const __m256 _prob_prev = _mm256_load_ps(&Alpha[(l-1)*modK+k]);
-					const __m256 _prob_temp = _mm256_fmadd_ps(_prob_prev, _fact2, _fact1);
-					const __m256 _prob_curr = _mm256_mul_ps(_prob_temp, _emiss);
-					_sum = _mm256_add_ps(_sum, _prob_curr);
-					_mm256_store_ps(&Alpha[l*modK+k], _prob_curr);
+					const simde__m256i _mask = simde_mm256_sllv_epi32(simde_mm256_set1_epi32((unsigned int )C->Hvar.getByte(l, k)), _vshift_count);
+					const simde__m256 _emiss = simde_mm256_blendv_ps (_emit0, _emit1, simde_mm256_castsi256_ps(_mask));
+					const simde__m256 _prob_prev = simde_mm256_load_ps(&Alpha[(l-1)*modK+k]);
+					const simde__m256 _prob_temp = simde_mm256_fmadd_ps(_prob_prev, _fact2, _fact1);
+					const simde__m256 _prob_curr = simde_mm256_mul_ps(_prob_temp, _emiss);
+					_sum = simde_mm256_add_ps(_sum, _prob_curr);
+					simde_mm256_store_ps(&Alpha[l*modK+k], _prob_curr);
 				}
 				if (k) AlphaSum[l] = horizontal_add(_sum);
 				for (int offset = nstatesMD8; offset < nstates ; offset ++)
@@ -175,39 +175,39 @@ void imputation_hmm::backward(const std::vector < float > & HL, std::vector < bo
 {
 	float betaSum = 0.0f, betaSumNext = 0.0f;
 	std::array<float,2> prob_hid, prob_obs;
-	const __m256i _vshift_count = _mm256_set_epi32(31,30,29,28,27,26,25,24);
+	const simde__m256i _vshift_count = simde_mm256_set_epi32(31,30,29,28,27,26,25,24);
 	const unsigned int nstates = C->n_states;
 	const unsigned int nstatesMD8 = (nstates / 8) * 8;
-	const __m256 _zero = _mm256_set1_ps(0.0f);
-	const __m256 _one = _mm256_set1_ps(1.0f);
+	const simde__m256 _zero = simde_mm256_set1_ps(0.0f);
+	const simde__m256 _one = simde_mm256_set1_ps(1.0f);
 	fill(Beta.begin(), Beta.end(), 1.0f);
-	__m256 _sum,  _prob0, _prob1;
+	simde__m256 _sum,  _prob0, _prob1;
 	for (int l = C->polymorphic_sites.size()-1 ; l >= 0 ; l --)
 	{
 		betaSum=0.0f;
 		prob_hid[0]=0.0f;
 		prob_hid[1]=0.0f;
 
-		_sum = _mm256_set1_ps(0.0f);
-		_prob0 = _mm256_set1_ps(0.0f);
-		_prob1 = _mm256_set1_ps(0.0f);
+		_sum = simde_mm256_set1_ps(0.0f);
+		_prob0 = simde_mm256_set1_ps(0.0f);
+		_prob1 = simde_mm256_set1_ps(0.0f);
 
 		if (flat[C->polymorphic_sites[l]] || C->lq_flag[C->polymorphic_sites[l]])
 		{
 			if (l == (C->polymorphic_sites.size()-1))
 			{
 				const float fact1 = 1.0f / nstates;
-				const __m256 _fact1 = _mm256_set1_ps(fact1);
+				const simde__m256 _fact1 = simde_mm256_set1_ps(fact1);
 				int k = 0;
 				for (; k < nstatesMD8; k += 8)
 				{
-					const __m256i _mask_curr = _mm256_sllv_epi32(_mm256_set1_epi32((unsigned int )C->Hvar.getByte(l, k)), _vshift_count);
-					const __m256 _mask0 = _mm256_blendv_ps (_one, _zero, _mm256_castsi256_ps(_mask_curr));
-					const __m256 _mask1 = _mm256_blendv_ps (_zero, _one, _mm256_castsi256_ps(_mask_curr));
-					const __m256 _alphas = _mm256_load_ps(&Alpha[l*modK+k]);
-					_prob0 = _mm256_add_ps(_prob0, _mm256_mul_ps(_alphas, _mask0));
-					_prob1 = _mm256_add_ps(_prob1, _mm256_mul_ps(_alphas, _mask1));
-					_mm256_store_ps(&Beta[k], _fact1);
+					const simde__m256i _mask_curr = simde_mm256_sllv_epi32(simde_mm256_set1_epi32((unsigned int )C->Hvar.getByte(l, k)), _vshift_count);
+					const simde__m256 _mask0 = simde_mm256_blendv_ps (_one, _zero, simde_mm256_castsi256_ps(_mask_curr));
+					const simde__m256 _mask1 = simde_mm256_blendv_ps (_zero, _one, simde_mm256_castsi256_ps(_mask_curr));
+					const simde__m256 _alphas = simde_mm256_load_ps(&Alpha[l*modK+k]);
+					_prob0 = simde_mm256_add_ps(_prob0, simde_mm256_mul_ps(_alphas, _mask0));
+					_prob1 = simde_mm256_add_ps(_prob1, simde_mm256_mul_ps(_alphas, _mask1));
+					simde_mm256_store_ps(&Beta[k], _fact1);
 				}
 				if (k)
 				{
@@ -225,23 +225,23 @@ void imputation_hmm::backward(const std::vector < float > & HL, std::vector < bo
 			{
 				const float fact1 = C->t[l] / nstates;
 				const float fact2 = C->nt[l] / betaSumNext;
-				const __m256 _fact1 = _mm256_set1_ps(fact1);
-				const __m256 _fact2 = _mm256_set1_ps(fact2);
+				const simde__m256 _fact1 = simde_mm256_set1_ps(fact1);
+				const simde__m256 _fact2 = simde_mm256_set1_ps(fact2);
 
 				int k = 0;
 				for (; k < nstatesMD8; k += 8)
 				{
-					const __m256i _mask_curr = _mm256_sllv_epi32(_mm256_set1_epi32((unsigned int )C->Hvar.getByte(l, k)), _vshift_count);
-					const __m256 _mask0 = _mm256_blendv_ps (_one, _zero, _mm256_castsi256_ps(_mask_curr));
-					const __m256 _mask1 = _mm256_blendv_ps (_zero, _one, _mm256_castsi256_ps(_mask_curr));
-					const __m256 _prob_prev = _mm256_load_ps(&Beta[k]);
-					const __m256 _prob_curr = _mm256_fmadd_ps(_prob_prev, _fact2, _fact1);
-					const __m256 _alphas = _mm256_load_ps(&Alpha[l*modK+k]);
-					const __m256 _dotprod = _mm256_mul_ps(_alphas, _prob_curr);
-					_prob0 = _mm256_add_ps(_prob0, _mm256_mul_ps(_dotprod, _mask0));
-					_prob1 = _mm256_add_ps(_prob1, _mm256_mul_ps(_dotprod, _mask1));
-					_sum = _mm256_add_ps(_sum, _prob_curr);
-					_mm256_store_ps(&Beta[k], _prob_curr);
+					const simde__m256i _mask_curr = simde_mm256_sllv_epi32(simde_mm256_set1_epi32((unsigned int )C->Hvar.getByte(l, k)), _vshift_count);
+					const simde__m256 _mask0 = simde_mm256_blendv_ps (_one, _zero, simde_mm256_castsi256_ps(_mask_curr));
+					const simde__m256 _mask1 = simde_mm256_blendv_ps (_zero, _one, simde_mm256_castsi256_ps(_mask_curr));
+					const simde__m256 _prob_prev = simde_mm256_load_ps(&Beta[k]);
+					const simde__m256 _prob_curr = simde_mm256_fmadd_ps(_prob_prev, _fact2, _fact1);
+					const simde__m256 _alphas = simde_mm256_load_ps(&Alpha[l*modK+k]);
+					const simde__m256 _dotprod = simde_mm256_mul_ps(_alphas, _prob_curr);
+					_prob0 = simde_mm256_add_ps(_prob0, simde_mm256_mul_ps(_dotprod, _mask0));
+					_prob1 = simde_mm256_add_ps(_prob1, simde_mm256_mul_ps(_dotprod, _mask1));
+					_sum = simde_mm256_add_ps(_sum, _prob_curr);
+					simde_mm256_store_ps(&Beta[k], _prob_curr);
 				}
 				if (k)
 				{
@@ -268,27 +268,27 @@ void imputation_hmm::backward(const std::vector < float > & HL, std::vector < bo
 		else
 		{
 			std::array<float,2> emit = {Emissions[2*C->polymorphic_sites[l]+0], Emissions[2*C->polymorphic_sites[l]+1]};
-			const __m256 _emit0 = _mm256_set1_ps(emit[0]);
-			const __m256 _emit1 = _mm256_set1_ps(emit[1]);
+			const simde__m256 _emit0 = simde_mm256_set1_ps(emit[0]);
+			const simde__m256 _emit1 = simde_mm256_set1_ps(emit[1]);
 
 			if (l == (C->polymorphic_sites.size()-1))
 			{
 				const float fact1 = 1.0f / nstates;
-				const __m256 _fact1 = _mm256_set1_ps(fact1);
+				const simde__m256 _fact1 = simde_mm256_set1_ps(fact1);
 
 				int k = 0;
 				for (; k < nstatesMD8; k += 8)
 				{
-					const __m256i _mask_curr = _mm256_sllv_epi32(_mm256_set1_epi32((unsigned int )C->Hvar.getByte(l, k)), _vshift_count);
-					const __m256 _emiss = _mm256_blendv_ps (_emit0, _emit1, _mm256_castsi256_ps(_mask_curr));
-					const __m256 _mask0 = _mm256_blendv_ps (_one, _zero, _mm256_castsi256_ps(_mask_curr));
-					const __m256 _mask1 = _mm256_blendv_ps (_zero, _one, _mm256_castsi256_ps(_mask_curr));
-					const __m256 _alphas = _mm256_load_ps(&Alpha[l*modK+k]);
-					const __m256 _prob_next = _mm256_mul_ps(_emiss, _fact1);
-					_prob0 = _mm256_add_ps(_prob0, _mm256_mul_ps(_alphas, _mask0));
-					_prob1 = _mm256_add_ps(_prob1, _mm256_mul_ps(_alphas, _mask1));
-					_sum= _mm256_add_ps(_sum, _prob_next);
-					_mm256_store_ps(&Beta[k], _prob_next);
+					const simde__m256i _mask_curr = simde_mm256_sllv_epi32(simde_mm256_set1_epi32((unsigned int )C->Hvar.getByte(l, k)), _vshift_count);
+					const simde__m256 _emiss = simde_mm256_blendv_ps (_emit0, _emit1, simde_mm256_castsi256_ps(_mask_curr));
+					const simde__m256 _mask0 = simde_mm256_blendv_ps (_one, _zero, simde_mm256_castsi256_ps(_mask_curr));
+					const simde__m256 _mask1 = simde_mm256_blendv_ps (_zero, _one, simde_mm256_castsi256_ps(_mask_curr));
+					const simde__m256 _alphas = simde_mm256_load_ps(&Alpha[l*modK+k]);
+					const simde__m256 _prob_next = simde_mm256_mul_ps(_emiss, _fact1);
+					_prob0 = simde_mm256_add_ps(_prob0, simde_mm256_mul_ps(_alphas, _mask0));
+					_prob1 = simde_mm256_add_ps(_prob1, simde_mm256_mul_ps(_alphas, _mask1));
+					_sum= simde_mm256_add_ps(_sum, _prob_next);
+					simde_mm256_store_ps(&Beta[k], _prob_next);
 				}
 				if (k)
 				{
@@ -307,25 +307,25 @@ void imputation_hmm::backward(const std::vector < float > & HL, std::vector < bo
 			{
 				const float fact1 = C->t[l] / nstates;
 				const float fact2 = C->nt[l] / betaSumNext;
-				const __m256 _fact1 = _mm256_set1_ps(fact1);
-				const __m256 _fact2 = _mm256_set1_ps(fact2);
+				const simde__m256 _fact1 = simde_mm256_set1_ps(fact1);
+				const simde__m256 _fact2 = simde_mm256_set1_ps(fact2);
 
 				int k = 0;
 				for (; k < nstatesMD8; k += 8)
 				{
-					const __m256i _mask_curr = _mm256_sllv_epi32(_mm256_set1_epi32((unsigned int )C->Hvar.getByte(l, k)), _vshift_count);
-					const __m256 _emiss = _mm256_blendv_ps (_emit0, _emit1, _mm256_castsi256_ps(_mask_curr));
-					const __m256 _mask0 = _mm256_blendv_ps (_one, _zero, _mm256_castsi256_ps(_mask_curr));
-					const __m256 _mask1 = _mm256_blendv_ps (_zero, _one, _mm256_castsi256_ps(_mask_curr));
-					const __m256 _prob_prev = _mm256_load_ps(&Beta[k]);
-					const __m256 _alphas = _mm256_load_ps(&Alpha[l*modK+k]);
-					const __m256 _prob_curr = _mm256_fmadd_ps(_prob_prev, _fact2, _fact1);
-					const __m256 _dotprod = _mm256_mul_ps(_alphas, _prob_curr);
-					const __m256 _prob_next = _mm256_mul_ps(_prob_curr, _emiss);
-					_prob0 = _mm256_add_ps(_prob0, _mm256_mul_ps(_dotprod, _mask0));
-					_prob1 = _mm256_add_ps(_prob1, _mm256_mul_ps(_dotprod, _mask1));
-					_sum= _mm256_add_ps(_sum, _prob_next);
-					_mm256_store_ps(&Beta[k], _prob_next);
+					const simde__m256i _mask_curr = simde_mm256_sllv_epi32(simde_mm256_set1_epi32((unsigned int )C->Hvar.getByte(l, k)), _vshift_count);
+					const simde__m256 _emiss = simde_mm256_blendv_ps (_emit0, _emit1, simde_mm256_castsi256_ps(_mask_curr));
+					const simde__m256 _mask0 = simde_mm256_blendv_ps (_one, _zero, simde_mm256_castsi256_ps(_mask_curr));
+					const simde__m256 _mask1 = simde_mm256_blendv_ps (_zero, _one, simde_mm256_castsi256_ps(_mask_curr));
+					const simde__m256 _prob_prev = simde_mm256_load_ps(&Beta[k]);
+					const simde__m256 _alphas = simde_mm256_load_ps(&Alpha[l*modK+k]);
+					const simde__m256 _prob_curr = simde_mm256_fmadd_ps(_prob_prev, _fact2, _fact1);
+					const simde__m256 _dotprod = simde_mm256_mul_ps(_alphas, _prob_curr);
+					const simde__m256 _prob_next = simde_mm256_mul_ps(_prob_curr, _emiss);
+					_prob0 = simde_mm256_add_ps(_prob0, simde_mm256_mul_ps(_dotprod, _mask0));
+					_prob1 = simde_mm256_add_ps(_prob1, simde_mm256_mul_ps(_dotprod, _mask1));
+					_sum= simde_mm256_add_ps(_sum, _prob_next);
+					simde_mm256_store_ps(&Beta[k], _prob_next);
 				}
 				if (k)
 				{
diff --git a/phase/src/models/imputation_hmm.h b/phase/src/models/imputation_hmm.h
index 4e5503e..a4f609d 100644
--- a/phase/src/models/imputation_hmm.h
+++ b/phase/src/models/imputation_hmm.h
@@ -28,7 +28,6 @@
 
 #include <utils/otools.h>
 #include <containers/conditioning_set.h>
-#define SIMDE_ENABLE_NATIVE_ALIASES
 #include <simde/x86/avx2.h>
 #include <simde/x86/fma.h>
 #include <boost/align/aligned_allocator.hpp>
diff --git a/phase/src/models/phasing_hmm.h b/phase/src/models/phasing_hmm.h
index bf32c3b..d3531e5 100644
--- a/phase/src/models/phasing_hmm.h
+++ b/phase/src/models/phasing_hmm.h
@@ -28,7 +28,6 @@
 
 #include <utils/otools.h>
 #include <containers/conditioning_set.h>
-#define SIMDE_ENABLE_NATIVE_ALIASES
 #include <simde/x86/avx2.h>
 #include <simde/x86/fma.h>
 #include <boost/align/aligned_allocator.hpp>
@@ -45,16 +44,16 @@ using aligned_vector32 = std::vector<T, boost::alignment::aligned_allocator < T,
 #define ALLELE(hap , pos) ((hap) & (1<<(pos)))
 
 inline
-float horizontal_add (const __m256& a) //@simo: implemented horizontal add with instructions in the 4B minimum
+float horizontal_add (const simde__m256& a) //@simo: implemented horizontal add with instructions in the 4B minimum
 {
-    __m128 vlow = _mm256_castps256_ps128(a);
-    __m128 vhigh = _mm256_extractf128_ps(a, 1); // high 128
-   vlow = _mm_add_ps(vlow, vhigh);     // add the low 128
-   __m128 shuf = _mm_movehdup_ps(vlow);        // broadcast elements 3,1 to 2,0
-   __m128 sums = _mm_add_ps(vlow, shuf);
-   shuf = _mm_movehl_ps(shuf, sums); // high half -> low half
-   sums = _mm_add_ss(sums, shuf);
-   return _mm_cvtss_f32(sums);
+    simde__m128 vlow = simde_mm256_castps256_ps128(a);
+    simde__m128 vhigh = simde_mm256_extractf128_ps(a, 1); // high 128
+   vlow = simde_mm_add_ps(vlow, vhigh);     // add the low 128
+   simde__m128 shuf = simde_mm_movehdup_ps(vlow);        // broadcast elements 3,1 to 2,0
+   simde__m128 sums = simde_mm_add_ps(vlow, shuf);
+   shuf = simde_mm_movehl_ps(shuf, sums); // high half -> low half
+   sums = simde_mm_add_ss(sums, shuf);
+   return simde_mm_cvtss_f32(sums);
 }
 
 class phasing_hmm {
@@ -139,8 +138,8 @@ public:
 inline
 void phasing_hmm::INIT_PEAK_HET(int curr_het)
 {
-	const std::array <__m256, 2 > emits = {_mm256_load_ps(&EMIT0[curr_het][0]),_mm256_load_ps(&EMIT1[curr_het][0])};
-    __m256 _sum = _mm256_set1_ps(0.0f);
+	const std::array <simde__m256, 2 > emits = {simde_mm256_load_ps(&EMIT0[curr_het][0]),simde_mm256_load_ps(&EMIT1[curr_het][0])};
+    simde__m256 _sum = simde_mm256_set1_ps(0.0f);
 	const unsigned int n_states = C->n_states;
 	const unsigned int n_states_full = (n_states / 8) * 8;
 	int k = 0, i = 0;
@@ -150,25 +149,25 @@ void phasing_hmm::INIT_PEAK_HET(int curr_het)
 		for (int b = 0 ; b < 8 ; ++b)
 		{
 			const bool ah = (byte >> (7 - b)) & 1;
-			_sum = _mm256_add_ps(_sum, emits[ah]);
-			_mm256_store_ps(&prob[i + b * HAP_NUMBER], emits[ah]);
+			_sum = simde_mm256_add_ps(_sum, emits[ah]);
+			simde_mm256_store_ps(&prob[i + b * HAP_NUMBER], emits[ah]);
 		}
 	}
 	for ( ; k < n_states ; ++k, i += HAP_NUMBER)
 	{
 		const bool ah = C->Hvar.get(curr_rel_locus, k);
-		_sum = _mm256_add_ps(_sum, emits[ah]);
-		_mm256_store_ps(&prob[i], emits[ah]);
+		_sum = simde_mm256_add_ps(_sum, emits[ah]);
+		simde_mm256_store_ps(&prob[i], emits[ah]);
 	}
-	_mm256_store_ps(&probSumH[0], _sum);
+	simde_mm256_store_ps(&probSumH[0], _sum);
 	probSumT = horizontal_add(_sum);
 }
 
 inline
 void phasing_hmm::INIT_PEAK_HOM(bool ag)
 {
-	const std::array <__m256, 2 > emits = {_mm256_set1_ps(1.0f),_mm256_set1_ps(C->ed_phs/C->ee_phs)};
-    __m256 _sum = _mm256_set1_ps(0.0f);
+	const std::array <simde__m256, 2 > emits = {simde_mm256_set1_ps(1.0f),simde_mm256_set1_ps(C->ed_phs/C->ee_phs)};
+    simde__m256 _sum = simde_mm256_set1_ps(0.0f);
 	const unsigned int n_states = C->n_states;
 	const unsigned int n_states_full = (n_states / 8) * 8;
 	int k = 0, i = 0;
@@ -178,17 +177,17 @@ void phasing_hmm::INIT_PEAK_HOM(bool ag)
 		for (int b = 0 ; b < 8 ; ++b)
 		{
 			const bool ag_ah = ((byte >> (7 - b)) & 1) != ag;
-			_sum = _mm256_add_ps(_sum, emits[ag_ah]);
-			_mm256_store_ps(&prob[i + b * HAP_NUMBER], emits[ag_ah]);
+			_sum = simde_mm256_add_ps(_sum, emits[ag_ah]);
+			simde_mm256_store_ps(&prob[i + b * HAP_NUMBER], emits[ag_ah]);
 		}
 	}
 	for ( ; k < n_states ; ++k, i += HAP_NUMBER)
 	{
 		const bool ag_ah = C->Hvar.get(curr_rel_locus, k)!=ag;
-		_sum = _mm256_add_ps(_sum, emits[ag_ah]);
-		_mm256_store_ps(&prob[i], emits[ag_ah]);
+		_sum = simde_mm256_add_ps(_sum, emits[ag_ah]);
+		simde_mm256_store_ps(&prob[i], emits[ag_ah]);
 	}
-	_mm256_store_ps(&probSumH[0], _sum);
+	simde_mm256_store_ps(&probSumH[0], _sum);
 	probSumT = horizontal_add(_sum);
 }
 
@@ -202,10 +201,10 @@ void phasing_hmm::INIT_FLAT_HET() {
 inline
 void phasing_hmm::RUN_PEAK_HET(int curr_het)
 {
-	const __m256 _tFreq = _mm256_mul_ps(_mm256_load_ps(&probSumH[0]), _mm256_set1_ps(yt / (C->n_states * probSumT)));
-	const __m256 _nt = _mm256_set1_ps(nt / probSumT);
-	const std::array <__m256, 2 > emits = {_mm256_load_ps(&EMIT0[curr_het][0]),_mm256_load_ps(&EMIT1[curr_het][0])};
-    __m256 _sum = _mm256_set1_ps(0.0f);
+	const simde__m256 _tFreq = simde_mm256_mul_ps(simde_mm256_load_ps(&probSumH[0]), simde_mm256_set1_ps(yt / (C->n_states * probSumT)));
+	const simde__m256 _nt = simde_mm256_set1_ps(nt / probSumT);
+	const std::array <simde__m256, 2 > emits = {simde_mm256_load_ps(&EMIT0[curr_het][0]),simde_mm256_load_ps(&EMIT1[curr_het][0])};
+    simde__m256 _sum = simde_mm256_set1_ps(0.0f);
 	const unsigned int n_states = C->n_states;
 	const unsigned int n_states_full = (n_states / 8) * 8;
 	int k = 0, i = 0;
@@ -215,29 +214,29 @@ void phasing_hmm::RUN_PEAK_HET(int curr_het)
 		for (int b = 0 ; b < 8 ; ++b)
 		{
 			const bool ah = (byte >> (7 - b)) & 1;
-			const __m256 _prob_curr = _mm256_mul_ps(_mm256_fmadd_ps(_mm256_load_ps(&prob[i + b * HAP_NUMBER]), _nt, _tFreq), emits[ah]);
-			_sum = _mm256_add_ps(_sum, _prob_curr);
-			_mm256_store_ps(&prob[i + b * HAP_NUMBER], _prob_curr);
+			const simde__m256 _prob_curr = simde_mm256_mul_ps(simde_mm256_fmadd_ps(simde_mm256_load_ps(&prob[i + b * HAP_NUMBER]), _nt, _tFreq), emits[ah]);
+			_sum = simde_mm256_add_ps(_sum, _prob_curr);
+			simde_mm256_store_ps(&prob[i + b * HAP_NUMBER], _prob_curr);
 		}
 	}
 	for ( ; k < n_states ; ++k, i += HAP_NUMBER)
 	{
 		const bool ah = C->Hvar.get(curr_rel_locus, k);
-		const __m256 _prob_curr = _mm256_mul_ps(_mm256_fmadd_ps(_mm256_load_ps(&prob[i]), _nt, _tFreq), emits[ah]);
-		_sum = _mm256_add_ps(_sum, _prob_curr);
-		_mm256_store_ps(&prob[i], _prob_curr);
+		const simde__m256 _prob_curr = simde_mm256_mul_ps(simde_mm256_fmadd_ps(simde_mm256_load_ps(&prob[i]), _nt, _tFreq), emits[ah]);
+		_sum = simde_mm256_add_ps(_sum, _prob_curr);
+		simde_mm256_store_ps(&prob[i], _prob_curr);
 	}
-	_mm256_store_ps(&probSumH[0], _sum);
+	simde_mm256_store_ps(&probSumH[0], _sum);
 	probSumT = horizontal_add(_sum);
 }
 
 inline
 void phasing_hmm::RUN_PEAK_HOM(bool ag)
 {
-	const __m256 _tFreq = _mm256_mul_ps(_mm256_load_ps(&probSumH[0]), _mm256_set1_ps(yt / (C->n_states * probSumT)));
-	const __m256 _nt = _mm256_set1_ps(nt / probSumT);
-    const __m256 _mism = _mm256_set1_ps(C->ed_phs/C->ee_phs);
-    __m256 _sum = _mm256_set1_ps(0.0f);
+	const simde__m256 _tFreq = simde_mm256_mul_ps(simde_mm256_load_ps(&probSumH[0]), simde_mm256_set1_ps(yt / (C->n_states * probSumT)));
+	const simde__m256 _nt = simde_mm256_set1_ps(nt / probSumT);
+    const simde__m256 _mism = simde_mm256_set1_ps(C->ed_phs/C->ee_phs);
+    simde__m256 _sum = simde_mm256_set1_ps(0.0f);
 	const unsigned int n_states = C->n_states;
 	const unsigned int n_states_full = (n_states / 8) * 8;
 	int k = 0, i = 0;
@@ -247,49 +246,49 @@ void phasing_hmm::RUN_PEAK_HOM(bool ag)
 		for (int b = 0 ; b < 8 ; ++b)
 		{
 			const bool ah = (byte >> (7 - b)) & 1;
-			const __m256 _prob_prev = _mm256_load_ps(&prob[i + b * HAP_NUMBER]);
-			__m256 _prob_curr = _mm256_fmadd_ps(_prob_prev, _nt, _tFreq);
-			if (ag!=ah) _prob_curr = _mm256_mul_ps(_prob_curr, _mism);
-			_sum = _mm256_add_ps(_sum, _prob_curr);
-			_mm256_store_ps(&prob[i + b * HAP_NUMBER], _prob_curr);
+			const simde__m256 _prob_prev = simde_mm256_load_ps(&prob[i + b * HAP_NUMBER]);
+			simde__m256 _prob_curr = simde_mm256_fmadd_ps(_prob_prev, _nt, _tFreq);
+			if (ag!=ah) _prob_curr = simde_mm256_mul_ps(_prob_curr, _mism);
+			_sum = simde_mm256_add_ps(_sum, _prob_curr);
+			simde_mm256_store_ps(&prob[i + b * HAP_NUMBER], _prob_curr);
 		}
 	}
 	for ( ; k < n_states ; ++k, i += HAP_NUMBER)
 	{
 		const bool ah = C->Hvar.get(curr_rel_locus, k);
-		const __m256 _prob_prev = _mm256_load_ps(&prob[i]);
-		__m256 _prob_curr = _mm256_fmadd_ps(_prob_prev, _nt, _tFreq);
-		if (ag!=ah) _prob_curr = _mm256_mul_ps(_prob_curr, _mism);
-		_sum = _mm256_add_ps(_sum, _prob_curr);
-		_mm256_store_ps(&prob[i], _prob_curr);
+		const simde__m256 _prob_prev = simde_mm256_load_ps(&prob[i]);
+		simde__m256 _prob_curr = simde_mm256_fmadd_ps(_prob_prev, _nt, _tFreq);
+		if (ag!=ah) _prob_curr = simde_mm256_mul_ps(_prob_curr, _mism);
+		_sum = simde_mm256_add_ps(_sum, _prob_curr);
+		simde_mm256_store_ps(&prob[i], _prob_curr);
 	}
-	_mm256_store_ps(&probSumH[0], _sum);
+	simde_mm256_store_ps(&probSumH[0], _sum);
 	probSumT = horizontal_add(_sum);
 }
 
 inline
 void phasing_hmm::RUN_FLAT_HET()
 {
-	const __m256 _tFreq = _mm256_mul_ps(_mm256_load_ps(&probSumH[0]), _mm256_set1_ps(yt / (C->n_states * probSumT)));
-	const __m256 _nt = _mm256_set1_ps(nt / probSumT);
-    __m256 _sum = _mm256_set1_ps(0.0f);
+	const simde__m256 _tFreq = simde_mm256_mul_ps(simde_mm256_load_ps(&probSumH[0]), simde_mm256_set1_ps(yt / (C->n_states * probSumT)));
+	const simde__m256 _nt = simde_mm256_set1_ps(nt / probSumT);
+    simde__m256 _sum = simde_mm256_set1_ps(0.0f);
 	for(int k = 0, i = 0 ; k != C->n_states ; ++k, i += HAP_NUMBER)
 	{
-		const __m256 _prob_curr = _mm256_fmadd_ps(_mm256_load_ps(&prob[i]), _nt, _tFreq);
-		_sum = _mm256_add_ps(_sum, _prob_curr);
-		_mm256_store_ps(&prob[i], _prob_curr);
+		const simde__m256 _prob_curr = simde_mm256_fmadd_ps(simde_mm256_load_ps(&prob[i]), _nt, _tFreq);
+		_sum = simde_mm256_add_ps(_sum, _prob_curr);
+		simde_mm256_store_ps(&prob[i], _prob_curr);
 	}
-	_mm256_store_ps(&probSumH[0], _sum);
+	simde_mm256_store_ps(&probSumH[0], _sum);
 	probSumT = horizontal_add(_sum);
 }
 
 inline
 void phasing_hmm::COLLAPSE_PEAK_HET(int curr_het)
 {
-	const __m256 _tFreq = _mm256_mul_ps(_mm256_load_ps(&probSumH[0]), _mm256_set1_ps(yt / (C->n_states * probSumT)));
-	const __m256 _nt = _mm256_set1_ps(nt / probSumT);
-	const std::array <__m256, 2 > emits = {_mm256_load_ps(&EMIT0[curr_het][0]),_mm256_load_ps(&EMIT1[curr_het][0])};
-    __m256 _sum = _mm256_set1_ps(0.0f);
+	const simde__m256 _tFreq = simde_mm256_mul_ps(simde_mm256_load_ps(&probSumH[0]), simde_mm256_set1_ps(yt / (C->n_states * probSumT)));
+	const simde__m256 _nt = simde_mm256_set1_ps(nt / probSumT);
+	const std::array <simde__m256, 2 > emits = {simde_mm256_load_ps(&EMIT0[curr_het][0]),simde_mm256_load_ps(&EMIT1[curr_het][0])};
+    simde__m256 _sum = simde_mm256_set1_ps(0.0f);
 	const unsigned int n_states = C->n_states;
 	const unsigned int n_states_full = (n_states / 8) * 8;
 	int k = 0, i = 0;
@@ -299,29 +298,29 @@ void phasing_hmm::COLLAPSE_PEAK_HET(int curr_het)
 		for (int b = 0 ; b < 8 ; ++b)
 		{
 			const bool ah = (byte >> (7 - b)) & 1;
-			const __m256 _prob_curr = _mm256_mul_ps(_mm256_fmadd_ps(_mm256_set1_ps(probSumK[k + b]), _nt, _tFreq), emits[ah]);
-			_sum = _mm256_add_ps(_sum, _prob_curr);
-			_mm256_store_ps(&prob[i + b * HAP_NUMBER], _prob_curr);
+			const simde__m256 _prob_curr = simde_mm256_mul_ps(simde_mm256_fmadd_ps(simde_mm256_set1_ps(probSumK[k + b]), _nt, _tFreq), emits[ah]);
+			_sum = simde_mm256_add_ps(_sum, _prob_curr);
+			simde_mm256_store_ps(&prob[i + b * HAP_NUMBER], _prob_curr);
 		}
 	}
 	for ( ; k < n_states ; ++k, i += HAP_NUMBER)
 	{
 		const bool ah = C->Hvar.get(curr_rel_locus, k);
-		const __m256 _prob_curr = _mm256_mul_ps(_mm256_fmadd_ps(_mm256_set1_ps(probSumK[k]), _nt, _tFreq), emits[ah]);
-		_sum = _mm256_add_ps(_sum, _prob_curr);
-		_mm256_store_ps(&prob[i], _prob_curr);
+		const simde__m256 _prob_curr = simde_mm256_mul_ps(simde_mm256_fmadd_ps(simde_mm256_set1_ps(probSumK[k]), _nt, _tFreq), emits[ah]);
+		_sum = simde_mm256_add_ps(_sum, _prob_curr);
+		simde_mm256_store_ps(&prob[i], _prob_curr);
 	}
-	_mm256_store_ps(&probSumH[0], _sum);
+	simde_mm256_store_ps(&probSumH[0], _sum);
 	probSumT = horizontal_add(_sum);
 }
 
 inline
 void phasing_hmm::COLLAPSE_PEAK_HOM(bool ag)
 {
-	const __m256 _tFreq = _mm256_mul_ps(_mm256_load_ps(&probSumH[0]), _mm256_set1_ps(yt / (C->n_states * probSumT)));
-   	const __m256 _nt = _mm256_set1_ps(nt / probSumT);
-    __m256 _sum = _mm256_set1_ps(0.0f);
-    const __m256 _mism = _mm256_set1_ps(C->ed_phs/C->ee_phs);
+	const simde__m256 _tFreq = simde_mm256_mul_ps(simde_mm256_load_ps(&probSumH[0]), simde_mm256_set1_ps(yt / (C->n_states * probSumT)));
+	const simde__m256 _nt = simde_mm256_set1_ps(nt / probSumT);
+	simde__m256 _sum = simde_mm256_set1_ps(0.0f);
+	const simde__m256 _mism = simde_mm256_set1_ps(C->ed_phs/C->ee_phs);
 	const unsigned int n_states = C->n_states;
 	const unsigned int n_states_full = (n_states / 8) * 8;
 	int k = 0, i = 0;
@@ -331,37 +330,37 @@ void phasing_hmm::COLLAPSE_PEAK_HOM(bool ag)
 		for (int b = 0 ; b < 8 ; ++b)
 		{
 			const bool ah = (byte >> (7 - b)) & 1;
-			__m256 _prob_curr = _mm256_fmadd_ps(_mm256_set1_ps(probSumK[k + b]), _nt, _tFreq);
-			if (ag!=ah) _prob_curr = _mm256_mul_ps(_prob_curr, _mism);
-			_sum = _mm256_add_ps(_sum, _prob_curr);
-			_mm256_store_ps(&prob[i + b * HAP_NUMBER], _prob_curr);
+			simde__m256 _prob_curr = simde_mm256_fmadd_ps(simde_mm256_set1_ps(probSumK[k + b]), _nt, _tFreq);
+			if (ag!=ah) _prob_curr = simde_mm256_mul_ps(_prob_curr, _mism);
+			_sum = simde_mm256_add_ps(_sum, _prob_curr);
+			simde_mm256_store_ps(&prob[i + b * HAP_NUMBER], _prob_curr);
 		}
 	}
-   	for ( ; k < n_states ; ++k, i += HAP_NUMBER)
-   	{
-   		const bool ah = C->Hvar.get(curr_rel_locus, k);
-   		__m256 _prob_curr = _mm256_fmadd_ps(_mm256_set1_ps(probSumK[k]), _nt, _tFreq);
-		if (ag!=ah) _prob_curr = _mm256_mul_ps(_prob_curr, _mism);
-   		_sum = _mm256_add_ps(_sum, _prob_curr);
-   		_mm256_store_ps(&prob[i], _prob_curr);
-   	}
-   	_mm256_store_ps(&probSumH[0], _sum);
-   	probSumT = horizontal_add(_sum);
+	for ( ; k < n_states ; ++k, i += HAP_NUMBER)
+	{
+		const bool ah = C->Hvar.get(curr_rel_locus, k);
+		simde__m256 _prob_curr = simde_mm256_fmadd_ps(simde_mm256_set1_ps(probSumK[k]), _nt, _tFreq);
+		if (ag!=ah) _prob_curr = simde_mm256_mul_ps(_prob_curr, _mism);
+		_sum = simde_mm256_add_ps(_sum, _prob_curr);
+		simde_mm256_store_ps(&prob[i], _prob_curr);
+	}
+	simde_mm256_store_ps(&probSumH[0], _sum);
+	probSumT = horizontal_add(_sum);
 }
 
 inline
 void phasing_hmm::COLLAPSE_FLAT_HET ()
 {
-	const __m256 _tFreq = _mm256_mul_ps(_mm256_load_ps(&probSumH[0]), _mm256_set1_ps(yt / (C->n_states * probSumT)));
-	const __m256 _nt = _mm256_set1_ps(nt / probSumT);
-    __m256 _sum = _mm256_set1_ps(0.0f);
+	const simde__m256 _tFreq = simde_mm256_mul_ps(simde_mm256_load_ps(&probSumH[0]), simde_mm256_set1_ps(yt / (C->n_states * probSumT)));
+	const simde__m256 _nt = simde_mm256_set1_ps(nt / probSumT);
+	simde__m256 _sum = simde_mm256_set1_ps(0.0f);
 	for(int k = 0, i = 0 ; k != C->n_states ; ++k, i += HAP_NUMBER)
 	{
-   		__m256 _prob_curr = _mm256_fmadd_ps(_mm256_set1_ps(probSumK[k]), _nt, _tFreq);
-		_sum = _mm256_add_ps(_sum, _prob_curr);
-		_mm256_store_ps(&prob[i], _prob_curr);
+		simde__m256 _prob_curr = simde_mm256_fmadd_ps(simde_mm256_set1_ps(probSumK[k]), _nt, _tFreq);
+		_sum = simde_mm256_add_ps(_sum, _prob_curr);
+		simde_mm256_store_ps(&prob[i], _prob_curr);
 	}
-	_mm256_store_ps(&probSumH[0], _sum);
+	simde_mm256_store_ps(&probSumH[0], _sum);
 	probSumT = horizontal_add(_sum);
 }
 
@@ -369,7 +368,7 @@ void phasing_hmm::COLLAPSE_FLAT_HET ()
 inline
 void phasing_hmm::SUMK() {
 	for(int k = 0, i = 0 ; k != C->n_states ; ++k, i += HAP_NUMBER)
-		probSumK[k] = horizontal_add(_mm256_load_ps(&prob[i]));
+		probSumK[k] = horizontal_add(simde_mm256_load_ps(&prob[i]));
 }
 
 inline
@@ -379,18 +378,18 @@ bool phasing_hmm::TRANS_HAP()
 	sumHProbs = 0.0f;
 	yt = C->getTransition(VAR_ABS[curr_idx_locus], VAR_ABS[curr_idx_locus+1]);
 	nt = 1.0f - yt;
-	const __m256 _fact2 = _mm256_set1_ps( nt / probSumT);
+	const simde__m256 _fact2 = simde_mm256_set1_ps( nt / probSumT);
 	int h1 = 0, j=0;
 	for (; h1 < HAP_NUMBER ; h1++, j += HAP_NUMBER)
 	{
-		const __m256 _fact1 = _mm256_set1_ps((probSumH[h1]/probSumT) * yt / C->n_states);
-		__m256 _sum = _mm256_set1_ps(0.0f);
+		const simde__m256 _fact1 = simde_mm256_set1_ps((probSumH[h1]/probSumT) * yt / C->n_states);
+		simde__m256 _sum = simde_mm256_set1_ps(0.0f);
 		for(int k=0, i=0; k != C->n_states ; ++k, i += HAP_NUMBER)
 		{
-			const __m256 _prob0 = _mm256_fmadd_ps(_mm256_set1_ps(prob[i+h1]), _fact2, _fact1);
-			_sum = _mm256_add_ps(_sum, _mm256_mul_ps(_prob0, _mm256_load_ps(&phasingProb[(curr_segment_index+1)*states_haps+i])));
+			const simde__m256 _prob0 = simde_mm256_fmadd_ps(simde_mm256_set1_ps(prob[i+h1]), _fact2, _fact1);
+			_sum = simde_mm256_add_ps(_sum, simde_mm256_mul_ps(_prob0, simde_mm256_load_ps(&phasingProb[(curr_segment_index+1)*states_haps+i])));
 		}
-		_mm256_store_ps(&HProbs[j], _sum);
+		simde_mm256_store_ps(&HProbs[j], _sum);
 		sumHProbs += horizontal_add(_sum);
 	}
 	return (std::isnan(sumHProbs) || std::isinf(sumHProbs) || sumHProbs < std::numeric_limits<float>::min());
@@ -414,13 +413,13 @@ inline
 void phasing_hmm::IMPUTE_FLAT_HET()
 {
 	const size_t states_haps = (size_t)C->n_states*HAP_NUMBER;
-	const __m256 _one = _mm256_set1_ps(1.0f);
-	const __m256 _zero = _mm256_set1_ps(0.0f);
-	__m256 _scaleR = _mm256_load_ps(&imputeProbSum[curr_missing_locus*HAP_NUMBER]);
-	__m256 _scaleL = _mm256_load_ps(&probSumH[0]);
-	_scaleR = _mm256_div_ps(_one, _scaleR);
-	_scaleL = _mm256_div_ps(_one, _scaleL);
-	std::array <__m256, 2 > sums = {_mm256_set1_ps(0.0f),_mm256_set1_ps(0.0f)};
+	const simde__m256 _one = simde_mm256_set1_ps(1.0f);
+	const simde__m256 _zero = simde_mm256_set1_ps(0.0f);
+	simde__m256 _scaleR = simde_mm256_load_ps(&imputeProbSum[curr_missing_locus*HAP_NUMBER]);
+	simde__m256 _scaleL = simde_mm256_load_ps(&probSumH[0]);
+	_scaleR = simde_mm256_div_ps(_one, _scaleR);
+	_scaleL = simde_mm256_div_ps(_one, _scaleL);
+	std::array <simde__m256, 2 > sums = {simde_mm256_set1_ps(0.0f),simde_mm256_set1_ps(0.0f)};
 	const unsigned int n_states = C->n_states;
 	const unsigned int n_states_full = (n_states / 8) * 8;
 	int k = 0, i = 0;
@@ -430,20 +429,20 @@ void phasing_hmm::IMPUTE_FLAT_HET()
 		for (int b = 0 ; b < 8 ; ++b)
 		{
 			const bool ah = (byte >> (7 - b)) & 1;
-			const __m256 _p1 = _mm256_mul_ps(_mm256_load_ps(&imputeProb[curr_missing_locus*states_haps + i + b * HAP_NUMBER]), _scaleR);
-			const __m256 _p2 = _mm256_mul_ps(_mm256_load_ps(&prob[i + b * HAP_NUMBER]), _scaleL);
-			sums[ah] = _mm256_add_ps(sums[ah], _mm256_mul_ps(_p1,_p2));
+			const simde__m256 _p1 = simde_mm256_mul_ps(simde_mm256_load_ps(&imputeProb[curr_missing_locus*states_haps + i + b * HAP_NUMBER]), _scaleR);
+			const simde__m256 _p2 = simde_mm256_mul_ps(simde_mm256_load_ps(&prob[i + b * HAP_NUMBER]), _scaleL);
+			sums[ah] = simde_mm256_add_ps(sums[ah], simde_mm256_mul_ps(_p1,_p2));
 		}
 	}
 	for ( ; k < n_states ; ++k, i += HAP_NUMBER) {
 		const bool ah = C->Hvar.get(curr_rel_locus, k);
-		const __m256 _p1 = _mm256_mul_ps(_mm256_load_ps(&imputeProb[curr_missing_locus*states_haps + i]), _scaleR);
-		const __m256 _p2 = _mm256_mul_ps(_mm256_load_ps(&prob[i]), _scaleL);
-		sums[ah] = _mm256_add_ps(sums[ah], _mm256_mul_ps(_p1,_p2));
+		const simde__m256 _p1 = simde_mm256_mul_ps(simde_mm256_load_ps(&imputeProb[curr_missing_locus*states_haps + i]), _scaleR);
+		const simde__m256 _p2 = simde_mm256_mul_ps(simde_mm256_load_ps(&prob[i]), _scaleL);
+		sums[ah] = simde_mm256_add_ps(sums[ah], simde_mm256_mul_ps(_p1,_p2));
 	}
-	const __m256 _norm_sum = _mm256_div_ps(sums[1], _mm256_add_ps(sums[0], sums[1]));
-	const __m256 _clamp = _mm256_max_ps(_zero, _mm256_min_ps(_one, _norm_sum));
-	_mm256_store_ps(&imputeProbOf1s[curr_missing_locus*HAP_NUMBER], _clamp);
+	const simde__m256 _norm_sum = simde_mm256_div_ps(sums[1], simde_mm256_add_ps(sums[0], sums[1]));
+	const simde__m256 _clamp = simde_mm256_max_ps(_zero, simde_mm256_min_ps(_one, _norm_sum));
+	simde_mm256_store_ps(&imputeProbOf1s[curr_missing_locus*HAP_NUMBER], _clamp);
 }
 
 #endif
